### PR TITLE
Support deleting membership when cluster is deleted

### PR DESCRIPTION
**Pull request by mnitchev**
Implements giantswarm/roadmap#1379

### CODE REVIEW

File: pkg/gke/membership/client.go [hunk 1]
```diff
@@ -60,6 +59,33 @@ func (c *Client) RegisterMembership(ctx context.Context, cluster *capg.GCPCluste
...
+	req := &gkehubpb.DeleteMembershipRequest{
+		Name: generateMembershipName(cluster),
+	}
+	op, err := c.gkeClient.DeleteMembership(ctx, req)
+	if hasHttpCode(err, http.StatusNotFound) {
+		logger.Info("membership already unregistered")
+		return nil
+	}
+	if err != nil {
+		logger.Error(err, "failed to create membership")
```
Comment (zewolfe) [inline pkg/gke/membership/client.go]:
```suggestion
		logger.Error(err, "failed to delete membership")
```

File: pkg/gke/membership/client.go [hunk 2]
```diff
@@ -60,6 +59,33 @@ func (c *Client) RegisterMembership(ctx context.Context, cluster *capg.GCPCluste
...
+		return err
+	}
+
+	err = op.Wait(ctx)
+	if err != nil {
+		logger.Error(err, "create membership operation failed")
```
Comment (zewolfe) [inline pkg/gke/membership/client.go]:
```suggestion
		logger.Error(err, "delete membership operation failed")
```

**Review by zewolfe — COMMENTED** [review]
Looks good. Just picked up these two.

**Review by zewolfe — APPROVED** [review]
👍🏾

**Comment by zewolfe** [comment]
We should also reference the issue for this PR

File: controllers/gcpcluster_controller.go [hunk 3]
```diff
@@ -43,28 +46,38 @@ const (
...
 	MembershipSecretNamespace string
-	GKEMembershipClient       GKEMembershipClient
+
+	runtimeClient       client.Client
+	GKEMembershipClient GKEMembershipClient
+}
+
+func NewGCPClusterReconciler(membershipSecretNamespace string, runtimeClient client.Client, membershipClient GKEMembershipClient) *GCPClusterReconciler {
+	return &GCPClusterReconciler{
+		runtimeClient:             runtimeClient,
```
Comment (fiunchinho) [inline controllers/gcpcluster_controller.go:63]: how come we need to store the client on a different field?

Comment (mnitchev) [reply controllers/gcpcluster_controller.go:63]: you mean instead of embedding it?

Comment (fiunchinho) [reply controllers/gcpcluster_controller.go:63]: Yes, like we had before.

File: controllers/gcpcluster_controller.go [hunk 4]
```diff
@@ -140,58 +187,73 @@ func (r *GCPClusterReconciler) Reconcile(ctx context.Context, req ctrl.Request)
...
+	return r.runtimeClient.Patch(ctx, cluster, client.MergeFrom(originalCluster))
+}
+
 func (r *GCPClusterReconciler) hasWorkloadIdentityEnabled(cluster *capg.GCPCluster) bool {
 	_, exists := cluster.Annotations[AnnotationWorkloadIdentityEnabled]
 	return exists
 }
 
-func (r *GCPClusterReconciler) getWorkloadClusterConfig(ctx context.Context, cluster *capg.GCPCluster, namespace string) (*rest.Config, error) {
+func (r *GCPClusterReconciler) getWorkloadClusterConfig(ctx context.Context, logger logr.Logger, cluster *capg.GCPCluster) (*rest.Config, error) {
```
Comment (fiunchinho) [inline controllers/gcpcluster_controller.go:207]: Could we take the logger from the `context` here?

File: controllers/gcpcluster_controller.go (hunk 3 above)
Comment (mnitchev) [reply controllers/gcpcluster_controller.go:63]: I just don't see the point of it. I don't want the reconciler to implement the `client.Client` interface, I just want it to use a client when reconciling

**Review by fiunchinho — APPROVED** [review]
Just the question about the logger, LGTM otherwise

File: controllers/gcpcluster_controller.go (hunk 4 above)
Comment (mnitchev) [reply controllers/gcpcluster_controller.go:207]: In this case we can yes, but I'll do it in the next PR

**Pull request merged**